### DISCUSSION TITLE
v5.7.1

### DISCUSSION THREAD

**Discussion by panva** [panva/openid-client#730] [Releases]
### Refactor

* jwt client authentication audience is now an issuer identifier string ([0b05217](https://github.com/panva/openid-client/commit/0b05217e7f283b75fd93c27c0f8c647f37501a33))

Note: If needed this can be reverted using the `extras.clientAssertionPayload` option.

<hr /><em>This discussion was created from the release <a href='https://github.com/panva/openid-client/releases/tag/v5.7.1'>v5.7.1</a>.</em>

**Comment by nwalters512** [comment]
@panva can you share more about the rationale for this change? Unfortunately this was very much a breaking change for us, and it's a shame to see it was released in a patch version.

From my read of the OIDC spec (https://openid.net/specs/openid-connect-core-1_0.html#rfc.section.9), under "9.  Client Authentication", "private_key_jwk", `aud` is indeed supposed to be the "Authorization Server's Token Endpoint", not the issuer. It'd be helpful to know which part of the spec this change was based on.

**Reply by panva** [reply]
> it's a shame to see it was released in a patch version

Unfortunately, it had to be done.

> It'd be helpful to know which part of the spec this change was based on.

A future not yet disclosed one.

> Unfortunately this was very much a breaking change for us

If needed this can be reverted using the `extras.clientAssertionPayload` option in whichever authenticated methods you invoke.

> From my read of the OIDC spec ([openid.net/specs/openid-connect-core-1_0.html#rfc.section.9](https://openid.net/specs/openid-connect-core-1_0.html#rfc.section.9)), under "9. Client Authentication", "private_key_jwk", aud is indeed supposed to be the "Authorization Server's Token Endpoint", not the issuer.

It's not that simple. The specs are layered in such a way that the issuer identifier is also a valid value for an AS to accept.

**Reply by nwalters512** [reply]
Sure, thankfully it was an easy enough fix (once we realized what was going wrong, which took a few days): https://github.com/PrairieLearn/PrairieLearn/pull/11038. I appreciate the pointer in the commit message.

Is this related to a vulnerability that hasn't been publicly disclosed yet? I can understand if you can't say more at this time, but otherwise the secrecy/caginess here is unexpected from this project.

Regardless, I appreciate the work you do in maintaining this library. Thank you.

**Reply by nwalters512** [reply]
> It's not that simple. The specs are layered in such a way that the issuer identifier is also a valid value for an AS to accept.

I buy that; I'm definitely not very well versed in the spec, this is just what I found with a quick read. For my own edification, do you know offhand what other pieces of the spec apply here?

**Reply by panva** [reply]
[I cannot say more at this time, the issue will be discussed publicly in the IETF OAuth WG interim meeting on 2025-01-27.](https://datatracker.ietf.org/wg/oauth/meetings/)

**Reply by nwalters512** [reply]
@panva are you able to share the relevant parts of the spec now? I'd like to forward them to platforms that my product works with so that they can update their implementation to accept their issuer string.

**Reply by panva** [reply]
The working group will work with this I-D as initial input for the updates

https://datatracker.ietf.org/doc/html/draft-jones-oauth-rfc7523bis-00

**Reply by panva** [reply]
Updated link to a WG document https://datatracker.ietf.org/doc/draft-ietf-oauth-rfc7523bis/
